File: cpp/quick_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

int partition(vector<int>& arr, int left, int right){
    int j = left;
    int pivot = arr[right];
    for(int i=left; i<right;i++){
        if(arr[i]<pivot) {
            swap(arr[i], arr[j]);
            j++;
        }
    }
    swap(arr[j], arr[right]);
    return j;
}

void quick_sort(vector<int>& arr, int left, int right){
    if(left>=right) return;
    int index = partition(arr, left, right);
    quick_sort(arr, left, index-1);
    quick_sort(arr, index+1, right);
}

int main(){
    vector<int> arr = {4,5,3,2,1,3};
    quick_sort(arr, 0, arr.size()-1);
    for(int i=0;i<arr.size();i++) cout<<arr[i]<<" ";
    cout<<endl;
    return 0;
}

File: java/sorts/quick_sort.java
import java.util.*;

public class quick_sort{
    public static void main(String[] args){
        quick_sort quick = new quick_sort();
        int[] arr = new int[]{15,13, 14, 1, 4};
        quick.sort(arr);
        for(int i=0;i<arr.length;i++) System.out.print(arr[i]+" ");
        System.out.println();
    }

    void sort(int[] arr){
        sortHelper(arr, 0, arr.length-1);
    }

    void sortHelper(int[] arr, int start, int end){
        if(start >= end) return;
        int pivot = partition(arr, start, end);
        sortHelper(arr, start, pivot-1);
        sortHelper(arr, pivot+1, end);
    }

    int partition(int[] arr, int start, int end){
        int pivot = arr[end];
        int i = start;
        int j = end-1;
        while(i<=j){
            while(i<=j && arr[i] < pivot) i++;
            while(i<=j && arr[j] >= pivot) j--;
            if(i<j) {
                swap(arr, i, j);
                i++;
                j--;
            }
        }
        arr[end] = arr[i];
        arr[i] = pivot;
        return i;
    }

    void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
